# 🔍 01 Data Exploration & Profiling Notebook
**โครงการ**: ETL & Data Warehouse สำหรับระบบวิเคราะห์รถยนต์มือสอง (Used Car Analytics)
**วัตถุประสงค์**: สำรวจโครงสร้างข้อมูล ตรวจสอบคุณภาพข้อมูล (Data Quality Audit) ทั้ง 3 แหล่งข้อมูลเพื่อจัดทำรายงานและเตรียมทำ ETL

---

In [ ]:
import pandas as pd
import numpy as np
import os

PATH_ONE2CAR = '../../01_Raw_Data/one2car/one2car_data.csv'
PATH_US_SALES = '../../01_Raw_Data/us-usecar/used_car_sales.csv'
PATH_SPEC1 = '../../01_Raw_Data/usecar-dataset/used_car_dataset.csv'
PATH_SPEC2 = '../../01_Raw_Data/usecar-dataset/used_cars_dataset_2.csv'

print('Libraries & File Paths Loaded Successfully!')

Libraries & File Paths Loaded Successfully!


## 🚗 1. Data Source 1: One2Car (ข้อมูลประกาศขายรถมือสองในไทย)
* **ที่มา**: `01_Raw_Data/one2car/one2car_data.csv` (รวมไฟล์ Web Scraped Data เรียบร้อยแล้ว)

In [ ]:
df_one2car = pd.read_csv(PATH_ONE2CAR)
print('One2car Shape:', df_one2car.shape[0], 'แถว,', df_one2car.shape[1], 'คอลัมน์')
df_one2car.head(5)

One2car Shape: 4190 แถว, 12 คอลัมน์


,web_scraper_order,web_scraper_start_url,pagination,car_title,price,description,mileage,location,car_model,condition,transmission,data4
0,1786479238-1,https://www.one2car.com/%E0%B8%A3%E0%B8%96%E0%...,NaN,2015 Honda City 1.5 (ปี 14-18) SV+ Sedan - SV,"269,000 บาท","HONDA CITY, 1.5 SV+ i-VTEC 2015 รถบ้านแท้มือเด...",170 - 175K กม.,กรุงเทพมหานคร,Honda City,มือสอง,เกียร์อัตโนมัติ,NaN
1,1786479238-2,https://www.one2car.com/%E0%B8%A3%E0%B8%96%E0%...,NaN,2023 Honda City 1.0 (ปี 19-26) SV Sedan,"359,000 บาท",'23 Honda City SV 1.0 Turbo ตัว Top รถมือเดียว...,30 - 35K กม.,สมุทรปราการ,Honda City,มือสอง,เกียร์อัตโนมัติ,NaN
2,1786479238-3,https://www.one2car.com/%E0%B8%A3%E0%B8%96%E0%...,NaN,2025 BMW 220i 2.0 F44 (ปี 20-27) Gran M Sport ...,"1,250,000 บาท",💢เจ้าของขายเองครับ💢 '25 แท้ BMW 220i M sport T...,30 - 35K กม.,สมุทรปราการ,NaN,มือสอง,เกียร์อัตโนมัติ,NaN
3,1786479238-4,https://www.one2car.com/%E0%B8%A3%E0%B8%96%E0%...,NaN,2022 Toyota HILUX REVO 2.4 Double Cab Z Editio...,"449,000 บาท",เกียร์ออโต้ ขับง่าย จดทะเบียนปี 2022 ( คศ 2021...,110 - 115K กม.,กรุงเทพมหานคร,Toyota HILUX REVO,มือสอง,เกียร์อัตโนมัติ,NaN
4,1786479238-5,https://www.one2car.com/%E0%B8%A3%E0%B8%96%E0%...,NaN,2015 Honda City 1.5 (ปี 14-18) SV Sedan,"259,000 บาท","ออกรถวันนี้ 1,000 บาท การันตีคุณภาพรถ ตรวจสภาพ...",20 - 25K กม.,กรุงเทพมหานคร,Honda City,มือสอง,เกียร์อัตโนมัติ,NaN


In [ ]:
# 1.2 ตรวจสอบค่าว่าง Missing Values & Duplicates
print('=== One2Car Missing Values ===')
print(df_one2car.isnull().sum())
print()
print('จำนวนแถวซ้ำทั้งหมด:', df_one2car.duplicated().sum(), 'แถว')

=== One2Car Missing Values ===
web_scraper_order           0
web_scraper_start_url       0
pagination                 94
car_title                   0
price                     223
description                 0
mileage                     0
location                  223
car_model                 751
condition                3271
transmission                0
data4                     919
dtype: int64

จำนวนแถวซ้ำทั้งหมด: 0 แถว


In [11]:
# 1.3 สำรวจรูปแบบข้อมูลคอลัมน์สำคัญ (price, mileage, location, transmission)
print('--- Sample Prices ---')
print(df_one2car['price'].dropna().unique()[:5])
print()
print('--- Sample Mileages ---')
print(df_one2car['mileage'].dropna().unique()[:5])
print()
print('--- Top Locations ---')
print(df_one2car['location'].value_counts().head(5))
print()
print('--- Transmission Distribution ---')
print(df_one2car['transmission'].value_counts())

--- Sample Prices ---
<StringArray>
['269,000 บาท', '359,000 บาท', '1,250,000 บาท', '449,000 บาท', '259,000 บาท']
Length: 5, dtype: str

--- Sample Mileages ---
<StringArray>
['170 - 175K กม.',   '30 - 35K กม.', '110 - 115K กม.',   '20 - 25K กม.',
   '10 - 15K กม.']
Length: 5, dtype: str

--- Top Locations ---
location
มือสอง           3048
กรุงเทพมหานคร     705
ชลบุรี             35
สมุทรปราการ        34
ปทุมธานี           29
Name: count, dtype: int64

--- Transmission Distribution ---
transmission
เกียร์อัตโนมัติ    3881
เกียร์ธรรมดา        309
Name: count, dtype: int64


## 🇺🇸 2. Data Source 2: US UseCar (ข้อมูลธุรกรรมการขายจริง)
* **ที่มา**: `01_Raw_Data/us-usecar/used_car_sales.csv`

In [12]:
# 2.1 โหลดข้อมูลธุรกรรมการขาย
df_us = pd.read_csv(PATH_US_SALES)
print('US Sales Shape:', df_us.shape[0], 'แถว,', df_us.shape[1], 'คอลัมน์')
df_us.head(5)

US Sales Shape: 122144 แถว, 13 คอลัมน์


,ID,pricesold,yearsold,zipcode,Mileage,Make,Model,Year,Trim,Engine,BodyType,NumCylinders,DriveType
0,137178,7500,2020,786**,84430,Ford,Mustang,1988,LX,5.0L Gas V8,Sedan,0,RWD
1,96705,15000,2019,81006,0,Replica/Kit Makes,Jaguar Beck Lister,1958,NaN,383 Fuel injected,Convertible,8,RWD
2,119660,8750,2020,33449,55000,Jaguar,XJS,1995,2+2 Cabriolet,4.0L In-Line 6 Cylinder,Convertible,6,RWD
3,80773,11600,2019,07852,97200,Ford,Mustang,1968,Stock,289 cu. in. V8,Coupe,8,RWD
4,64287,44000,2019,07728,40703,Porsche,911,2002,Turbo X-50,3.6L,Coupe,6,AWD


In [15]:
# 2.2 ตรวจสอบค่าว่าง สถิติราคาขาย และเลขไมล์
print('=== US Sales Missing Values ===')
print(df_us.isnull().sum())
print()
print('=== Summary Statistics (pricesold, yearsold, Mileage, Year) ===')
df_us[['pricesold', 'yearsold', 'Mileage', 'Year']].describe()

=== US Sales Missing Values ===
ID                  0
pricesold           0
yearsold            0
zipcode           909
Mileage             0
Make                0
Model             573
Year                0
Trim            48936
Engine          27204
BodyType        20786
NumCylinders        0
DriveType       24857
dtype: int64

=== Summary Statistics (pricesold, yearsold, Mileage, Year) ===


,pricesold,yearsold,Mileage,Year
count,122144.000000,122144.000000,1.221440e+05,1.221440e+05
mean,10808.560715,2019.375467,1.404291e+06,3.959362e+03
std,13987.295760,0.503671,3.335593e+07,1.984514e+05
min,0.000000,2018.000000,0.000000e+00,0.000000e+00
25%,2950.000000,2019.000000,4.479225e+04,1.977000e+03
50%,6500.000000,2019.000000,9.000000e+04,2.000000e+03
75%,13800.000000,2020.000000,1.402000e+05,2.008000e+03
max,404990.000000,2020.000000,1.235669e+09,2.014000e+07


## 🇮🇳 3. Data Source 3: UseCar Spec Dataset (ข้อมูลสเปกและคุณลักษณะเพิ่มเติม)
* **ที่มา**: รวมไฟล์ `used_car_dataset.csv` และ `used_cars_dataset_2.csv` ใน `01_Raw_Data/usecar-dataset/`

In [16]:
# 3.1 อ่านไฟล์และนำมารวมกัน (Consolidate Files)
df_spec1 = pd.read_csv(PATH_SPEC1)
df_spec2 = pd.read_csv(PATH_SPEC2)
df_spec_combined = pd.concat([df_spec1, df_spec2], ignore_index=True)

print('Spec Combined Shape:', df_spec_combined.shape[0], 'แถว,', df_spec_combined.shape[1], 'คอลัมน์')
df_spec_combined.head(5)

Spec Combined Shape: 24575 แถว, 11 คอลัมน์


,Brand,model,Year,Age,kmDriven,Transmission,Owner,FuelType,PostedDate,AdditionInfo,AskPrice
0,Honda,City,2001,23,"98,000 km",Manual,second,Petrol,Nov-24,"Honda City v teck in mint condition, valid gen...","₹ 1,95,000"
1,Toyota,Innova,2009,15,190000.0 km,Manual,second,Diesel,Jul-24,"Toyota Innova 2.5 G (Diesel) 7 Seater, 2009, D...","₹ 3,75,000"
2,Volkswagen,VentoTest,2010,14,"77,246 km",Manual,first,Diesel,Nov-24,"Volkswagen Vento 2010-2013 Diesel Breeze, 2010...","₹ 1,84,999"
3,Maruti Suzuki,Swift,2017,7,"83,500 km",Manual,second,Diesel,Nov-24,Maruti Suzuki Swift 2017 Diesel Good Condition,"₹ 5,65,000"
4,Maruti Suzuki,Baleno,2019,5,"45,000 km",Automatic,first,Petrol,Nov-24,"Maruti Suzuki Baleno Alpha CVT, 2019, Petrol","₹ 6,85,000"


In [17]:
# 3.2 ตรวจสอบค่าว่าง การกระจายตัวของ FuelType และ Owner
print('=== Spec Combined Missing Values ===')
print(df_spec_combined.isnull().sum())
print()
print('--- Fuel Types Distribution ---')
print(df_spec_combined['FuelType'].value_counts())
print()
print('--- Owner History Distribution ---')
print(df_spec_combined['Owner'].value_counts())

=== Spec Combined Missing Values ===
Brand             0
model             0
Year              0
Age               0
kmDriven        135
Transmission      0
Owner             0
FuelType          0
PostedDate        0
AdditionInfo      0
AskPrice          0
dtype: int64

--- Fuel Types Distribution ---
FuelType
Petrol        9644
Diesel        9079
Hybrid/CNG    3840
hybrid        2012
Name: count, dtype: int64

--- Owner History Distribution ---
Owner
second    12576
first     11999
Name: count, dtype: int64


--- 
## 📋 4. สรุปปัญหาคุณภาพข้อมูล 5 ประเภท (Data Quality Audit Summary)
*(สำหรับใช้เขียนอธิบายในรายงาน PDF หน้า 4-5 และตอบเกณฑ์อาจารย์)*

1. **Missing Values (ค่าว่าง)**: 
   - One2car: มีค่าว่างใน `price` (223 แถว), `location` (223 แถว), `car_model` (751 แถว)
   - US Sales: มีค่าว่างใน `BodyType` และ `Trim`
2. **Unit & Currency Mismatch (หน่วยและสกุลเงินต่างกัน)**: 
   - One2car ใช้ราคาภาษาไทยติดคำว่า "บาท" และไมล์ติดคำว่า "กม."
   - US Data ใช้เงินดอลลาร์ ($)
   - Spec Data ใช้เงินรูปี ("₹") และไมล์ติดคำว่า "km"
3. **Inconsistent Category Names (ชื่อหมวดหมู่ไม่สอดคล้องกัน)**:
   - ระบบเกียร์: "เกียร์อัตโนมัติ" vs "Automatic" / "เกียร์ธรรมดา" vs "Manual"
   - ชื่อยี่ห้อรถ: "car_model" vs "Make" vs "Brand"
4. **Invalid Values / Outliers (ค่าผิดปกติ)**:
   - US Dataset มีรายการที่ `pricesold` หรือ `Mileage` เป็น 0
5. **Mismatched Keys / Unstructured Text (ไม่มี Primary Key เชื่อมกัน)**:
   - ข้อมูล One2car ไม่มีคอลัมน์แยก `Year`, `Brand`, `Model` ที่ชัดเจน ต้องสกัดด้วย Regex จากข้อความ `car_title`